oauthenticator/azuread.py

```diff
@@ -0,0 +1,139 @@
+"""
+Custom Authenticator to use Azure AD with JupyterHub
+
+"""
+
+import json
+import jwt
+import os
+import re
+import string
+import urllib
+import sys
+
+from tornado.auth import OAuth2Mixin
+from tornado.log import app_log
+from tornado import gen, web
+
+from tornado.httputil import url_concat
+from tornado.httpclient import HTTPRequest, AsyncHTTPClient
+
+from jupyterhub.auth import LocalAuthenticator
+
+from traitlets import List, Set, Unicode
+
+from .common import next_page_from_links
+from .oauth2 import OAuthLoginHandler, OAuthenticator
+
+
+def _api_headers(access_token):
+    return {
+        "Accept": "application/json",
+        "User-Agent": "JupyterHub",
+        "Authorization": "token {}".format(access_token)
+    }
+
+
+# we need this, as microsoft expects '&' not to be encoded in the POST body
+def dictToQuery(d):
+    query = ''
+    for key in d.keys():
+        query += str(key) + '=' + str(d[key]) + "&"
+    return query
+
+
+def azureTokenUrlFor(tentant):
+    return 'https://login.microsoftonline.com/{0}/oauth2/token'.format(tentant)
+
+
+def azureAuthorizeUrlFor(tentant):
+    return 'https://login.microsoftonline.com/{0}/oauth2/authorize'.format(
+        tentant)
+
+
+class AzureAdMixin(OAuth2Mixin):
+    tenant_id = os.environ.get('AAD_TENANT_ID', '')
+    _OAUTH_ACCESS_TOKEN_URL = azureTokenUrlFor(tenant_id)
+    _OAUTH_AUTHORIZE_URL = azureAuthorizeUrlFor(tenant_id)
+
+
+class AzureAdLoginHandler(OAuthLoginHandler, AzureAdMixin):
+    pass
+
+
+class AzureAdOAuthenticator(OAuthenticator):
+    login_service = "Azure AD"
+
+    login_handler = AzureAdLoginHandler
+
+    tenant_id = Unicode(config=True)
+
+    def getTenant(self):
+        if hasattr(self, 'tenant_id') and self.tenant_id:
+            app_log.info('ID3: {0}'.format(self.tenant_id))
+            return self.tenant_id
+        else:
+            tenant_id = Unicode(
+                os.environ.get('AAD_TENANT_ID', ''),
+                config=True,
+                help="Tenant")
+            app_log.info('ID4: {0}'.format(tenant_id))
+            return tenant_id
+
+    @gen.coroutine
+    def authenticate(self, handler, data=None):
+        app_log.info('AZUREAD: authenticate')
+        """We set up auth_state based on additional GitHub info if we
```

Comment:
is this left over copy pasta from the github example? If so, can we nuke it?

Reply:
thanks!